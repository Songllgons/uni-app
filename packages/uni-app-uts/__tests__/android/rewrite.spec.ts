import MagicString from 'magic-string'
import { parse } from '@babel/parser'

import { rewriteConsole } from '../../src/plugins/android/uvue/sfc/compiler/script/rewriteConsole'
import { rewriteDebugError } from '../../src/plugins/android/uvue/sfc/compiler/script/rewriteDebugError'

function assertRewrite(input: string, startLine = 0, startOffset = 0): void {
  const ast = parse(input, {
    plugins: ['typescript'],
    sourceType: 'module',
  }).program
  const s = new MagicString(input)
  const options = {
    fileName: `test.uvue`,
    startLine,
    startOffset,
  }
  rewriteDebugError(ast, s, options)
  rewriteConsole(ast, s, options)
  expect(s.toString()).toMatchSnapshot()
}

describe('rewrite', () => {
  test(`console`, () => {
    assertRewrite(`console.log(1)`)
    assertRewrite(`

console.log()
`)
  })
  assertRewrite(`UTSAndroid.consoleDebugError(JSON.parse('{"a":1}'))`)
  assertRewrite(`
// 变量
const msg = 'Hello!'

// 函数
function log() {
    console.log(msg)
}  
  `)

  test(`JSON.parse`, () => {
    assertRewrite(
      `JSON.parse('{"a":1}');JSON.parseObject('{"a":1}');JSON.parseArray('[1,2,3]')`
    )
    assertRewrite(
      `JSON.parse('{"a":1}');
JSON.parseObject('{"a":1}');
JSON.parseArray('[1,2,3]')`
    )
  })
  test(`decode/encode`, () => {
    assertRewrite(
      `
decodeURI('')
decodeURIComponent('')
encodeURI('')
encodeURIComponent('')
`
    )
  })
})
