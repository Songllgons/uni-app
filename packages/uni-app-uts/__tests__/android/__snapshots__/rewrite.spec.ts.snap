// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `"UTSAndroid.consoleDebugError(UTSAndroid.consoleDebugError(JSON.parse('{"a":1}'), " at test.uvue:1"), " at test.uvue:1")"`;

exports[` 2`] = `
"
// 变量
const msg = 'Hello!'

// 函数
function log() {
    console.log(msg, " at test.uvue:7")
}  
  "
`;

exports[`rewrite JSON.parse 1`] = `"UTSAndroid.consoleDebugError(JSON.parse('{"a":1}'), " at test.uvue:1");UTSAndroid.consoleDebugError(JSON.parseObject('{"a":1}'), " at test.uvue:1");UTSAndroid.consoleDebugError(JSON.parseArray('[1,2,3]'), " at test.uvue:1")"`;

exports[`rewrite JSON.parse 2`] = `
"UTSAndroid.consoleDebugError(JSON.parse('{"a":1}'), " at test.uvue:1");
UTSAndroid.consoleDebugError(JSON.parseObject('{"a":1}'), " at test.uvue:2");
UTSAndroid.consoleDebugError(JSON.parseArray('[1,2,3]'), " at test.uvue:3")"
`;

exports[`rewrite console 1`] = `"console.log(1, " at test.uvue:1")"`;

exports[`rewrite console 2`] = `
"

console.log("at test.uvue:3")
"
`;

exports[`rewrite decode/encode 1`] = `
"
UTSAndroid.consoleDebugError(decodeURI(''), " at test.uvue:2")
UTSAndroid.consoleDebugError(decodeURIComponent(''), " at test.uvue:3")
UTSAndroid.consoleDebugError(encodeURI(''), " at test.uvue:4")
UTSAndroid.consoleDebugError(encodeURIComponent(''), " at test.uvue:5")
"
`;
